test/unit/api/solver_black.h

```diff
@@ -911,6 +892,73 @@ void SolverBlack::testDefineFunsRec()
       CVC4ApiException&);
 }
 
+void SolverBlack::testUFIteration()
+{
+  Sort intSort = d_solver->getIntegerSort();
+  Sort funSort = d_solver->mkFunctionSort({intSort, intSort}, intSort);
+  Term x = d_solver->mkConst(intSort, "x");
+  Term y = d_solver->mkConst(intSort, "y");
+  Term f = d_solver->mkConst(funSort, "f");
+  Term fxy = d_solver->mkTerm(APPLY_UF, f, x, y);
```

Comment:
Shouldn't this be `mkTerm(Op(APPLY_UF, f), x, y )` ?

Reply:
Not anymore, no. In the previous API, it would be more like that. Part of this refactor is to treat functions just as terms -- more of HOL perspective. One nice side effect of this, is that if you're traversing a formula and rebuilding it, you don't need to special-case functions. You just call `mkTerm` with `APPLY_UF` and the children.

Reply:
Ah, I understand now.

APPLY_UF might be a misnomer then, as `f` could be an arbitrary term whose type is a function.  Should we change this to APPLY?

Reply:
Ahh that's an interesting point! It would still map to the internal kind `CVC4::Kind::APPLY_UF` though. What other kinds of functions would you expect? Can you define functions through the API?

Reply:
Okay, so we do have `defineFunction` in the API. So that’s a good point. Under the hood, would you still use `APPLY_UF` for a defined function? My understanding was that it’s essentially a macro in CVC4, but I’m not sure where the substitution happens. Also, are there other cases besides UF and defined functions?

Reply:
A defined function (via define-fun) I would still consider to be an uninterpreted function for uniformity.

However, when we do higher-order SMT, three cases of function applications whose heads are not uninterpreted functions that come to mind are:
(1) A lambda expression, e.g. `(APPLY (lambda ((x Int)) (+ x 1)) c)`,
(2) A term ITE, e.g. `(APPLY (ite B f g) c)`,
(3) A universal variable e.g. `(APPLY x c)` in `(forall ((x (-> Int Bool))) (APPLY x c))`.

Reply:
Ahh I see, thanks for the explanation! Changing it to APPLY makes sense to me. But maybe we should do it in a separate PR? Seems slightly out of the scope of this one to me -- because this changes the semantics of term iteration and how OpTerms are handled. This APPLY_UF confusion already exists in master. Would you recommend renaming the internal kind as well? It seems strange to have `APPLY` map to `APPLY_UF` here:
https://github.com/CVC4/CVC4/blob/cbd86eb4ed8bafc17f28244b746a376a019462f1/src/api/cvc4cpp.cpp#L90

Reply:
Yes, I think it may be out of scope for this PR.